plugins/hls-refine-imports-plugin/src/Ide/Plugin/RefineImports.hs

```diff
@@ -0,0 +1,271 @@
+
+  let filterByImport
+        :: LImportDecl GhcRn
+        -> Map.Map ModuleName [AvailInfo]
+        -> Map.Map ModuleName [AvailInfo]
+      filterByImport (L _ ImportDecl{ideclHiding = Just (_, L _ names)}) avails =
+        let importedNames = map (prettyPrint . ieName . unLoc) names
+        in flip Map.filter avails $ \a ->
+              any ((`elem` importedNames) . prettyPrint)
+                $ concatMap availNamesWithSelectors a
+      filterByImport _ _ = mempty
+  let constructImport
+        :: LImportDecl GhcRn
+        -> (ModuleName, [AvailInfo])
+        -> LImportDecl GhcRn
+      constructImport
+        i@(L lim id@ImportDecl
+                  {ideclName = L _ mn, ideclHiding = Just (hiding, L _ names)})
+        (newModuleName, avails) = L lim id
+          { ideclName = noLoc newModuleName
+          , ideclHiding = Just (hiding, noLoc newNames)
+          }
+          where newNames = filter (\n -> any (n `containsAvail`) avails) names
+      constructImport lim _ = lim
+  let res =
+        [ (i, Just
+                . T.intercalate "\n"
+                . map (T.pack . prettyPrint . constructImport i)
+                . Map.toList
+                $ filteredInnerImports)
+        -- for every minimal imports
+        | Just minImports <- [mbMinImports]
+        , i@(L _ ImportDecl{ideclName = L _ mn}) <- minImports
+        -- we check for the inner imports
+        , Just innerImports <- [Map.lookup mn import2Map]
+        -- and only get those symbols used
+        , filteredInnerImports <- [filterByImport i innerImports]
+        -- if no symbols from this modules then don't need to generate new import
+        , not $ null filteredInnerImports
+        ]
+  return ([], RefineImportsResult res <$ mbMinImports)
+
+  where
+    -- Check if a name is exposed by AvailInfo (the available information of a module)
+    containsAvail :: LIE GhcRn -> AvailInfo -> Bool
+    containsAvail name avail =
+      any (\an -> prettyPrint an == (prettyPrint . ieName . unLoc $ name))
+        $ availNamesWithSelectors avail
+
+--------------------------------------------------------------------------------
+
+mkExplicitEdit :: PositionMapping -> LImportDecl pass -> T.Text -> Maybe TextEdit
+mkExplicitEdit posMapping (L src imp) explicit
+  | RealSrcSpan l <- src,
+    L _ mn <- ideclName imp,
+    -- (almost) no one wants to see an refine import list for Prelude
+    mn /= moduleName pRELUDE,
+    Just rng <- toCurrentRange posMapping $ realSrcSpanToRange l =
+    Just $ TextEdit rng explicit
+  | otherwise =
+    Nothing
+
+-- | Given an import declaration, generate a code lens unless it has an
+-- explicit import list or it's qualified
+generateLens :: PluginId -> Uri -> TextEdit -> IO (Maybe CodeLens)
+generateLens pId uri edits@TextEdit {_range, _newText} = do
+  -- The title of the command is just the minimal explicit import decl
+  let title = T.intercalate ", " (T.lines _newText)
```

Comment:
I think the title should be "Refine import to <new import>" since many users will be unaware of this feature